esy-native-install-npm-release/EsyNativeInstallNpmRelease.re

```diff
@@ -0,0 +1,214 @@
+    };
+  };
+  let initStore = () => {
+    Lwt_result.(
+      Fs.createDir(storePath)
+      >>= (
+        _ =>
+          RunAsync.List.waitAll([
+            Fs.createDir(Path.(storePath / storeBuildTree)),
+            Fs.createDir(Path.(storePath / storeInstallTree)),
+            Fs.createDir(Path.(storePath / storeStageTree)),
+          ])
+      )
+      |> Lwt_result.map_err(err => `EsyLibError(err))
+    );
+  };
+  let doImport = () => {
+    open RunAsync.Syntax;
+    let importBuilds = () => {
+      open RunAsync.Syntax;
+      let%bind files = Fs.listDir(releaseExportPath);
+
+      RunAsync.List.mapAndJoin(
+        ~f=file => importBuild(file, rewritePrefix),
+        files |> List.map(~f=Path.addSeg(releaseExportPath)),
+      );
+    };
+    let rewriteBinWrappers = () =>
+      switch (rewritePrefix) {
+      | NoRewrite(_) => Lwt_result.return()
+      | Rewrite(storePath) =>
+        let%bind prevStorePath =
+          Fs.readFile(Path.(releaseBinPath / "_storePath"));
+        RewritePrefix.rewritePrefix(
+          ~origPrefix=Path.v(prevStorePath),
+          ~destPrefix=storePath,
+          releaseBinPath,
+        );
+      };
+    let%bind _ = importBuilds();
+    rewriteBinWrappers();
+  };
+  Lwt_result.(
+    check()
+    >>= initStore
+    >>= (_ => doImport() |> Lwt_result.map_err(err => `EsyLibError(err)))
+    >>= (
+      _ => {
+        switch (System.Platform.host, System.Arch.host) {
+        | (Darwin, Arm64) =>
+          print_endline("Detected macOS arm64. Signing binaries...");
+          Result.Syntax.Let_syntax.bind(
+            EsyBuildPackage.Build.getMachOBins(
+              (module EsyBuildPackage.Run),
+              [],
+              Path.v(releasePackagePath),
+            ),
+            ~f=entries =>
+            EsyBuildPackage.BigSurArm.sign(entries)
+          )
+          |> Lwt.return;
+        | _ => Lwt.return(Ok())
+        };
+      }
+    )
+  );
+};
+
+let lwt_main = (ocamlPkgName, ocamlVersion, shouldRewritePrefix) => {
+  let unpaddedStorePath = Path.(v(releasePackagePath) / esyStoreVersion);
+
+  let rewritePrefixResult =
+    shouldRewritePrefix
+      ? getStorePathForPrefix(releasePackagePath, ocamlPkgName, ocamlVersion)
+        |> Result.map(~f=storePath => Rewrite(storePath))
+      : Ok(NoRewrite(unpaddedStorePath));
+
+  let result =
+    rewritePrefixResult
+    |> Result.Syntax.Let_syntax.bind(~f=rewritePrefix => {
+         Lwt_main.run(main(ocamlPkgName, ocamlVersion, rewritePrefix))
+       });
+  switch (result) {
+  | Ok(_) => print_endline("Done!")
+  | Error(`NoBuildFound) => Printf.eprintf("No build found!")
+  | Error(`ReleaseAlreadyInstalled) =>
+    Printf.eprintf("Release already installed!")
+  | Error(`Msg(msg)) => Printf.eprintf("%s", msg)
+  | Error(`EsyLibError(err)) =>
+    Printf.eprintf("%s", EsyLib.Run.formatError(err))
+  | Error(`CommandError(cmd, status)) =>
+    Bos.Cmd.pp(Format.err_formatter, cmd)
+  };
```

Comment:
I wanted to have the `main` function to return a `result` type with the `error` part being all variants so we could pattern match on them and decide what to do. Since we actually do nothing more than printing the error in every case, I might have made this more complicated than needed.

Happy to refactor and use the `RunAsync.error(string)` function to report these errors.

Reply:
Yes, that would be great! As much code/pattern reuse as possible.